examples/gno.land/r/demo/auction/auction.gno

```diff
@@ -0,0 +1,134 @@
+package auction
+
+import (
+	"bytes"
+	"std"
+	"strconv"
+	"time"
+
+	auctionp "gno.land/p/demo/auction"
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/ufmt"
+)
+
+var (
+	auctionTree *avl.Tree
+	currentTime time.Time
+)
+
+// Initialize the realm with auction tree
+func init() {
+	auctionTree = avl.NewTree()
+}
+
+// Mockable function to get current time
+func now() time.Time {
+	if !currentTime.IsZero() {
+		return currentTime
+	}
+	return time.Now()
+}
+
+// NewAuction creates a new auction in the realm
+func NewAuction(
+	title string,
+	description string,
+	begin int64,
+	end int64,
+	price uint64,
+) *auctionp.Auction {
+	txSender := std.GetOrigCaller()
+
+	if begin < now().Unix() {
+		panic("begin has to be in the future")
+	}
+	if end <= now().Unix() {
+		panic("end has to be in the future")
+	}
+	if end <= begin {
+		panic("end has to be after begin")
+	}
+	if price <= 0 {
+		panic("price has to be positive")
+	}
+
+	auc := auctionp.NewAuction(
+		title,
+		txSender,
+		description,
+		time.Unix(begin, 0),
+		time.Unix(end, 0),
+		price,
+	)
+	id := strconv.Itoa(auctionTree.Size())
```

Comment:
You can use the `seqid` package for sequential IDs, perfect for this AVL use case